src/Database/LSMTree/Internal/Run/Reader.hs

```diff
@@ -0,0 +1,165 @@
+    , readerRun            :: !(Run.Run fhandle)
+    }
+
+new ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> Run.Run (FS.Handle h)
+  -> IO (Reader (FS.Handle h))
+new fs bfs readerRun = do
+    readerKOpsHandle <-
+      FS.hOpen fs (runKOpsPath (Run.lsmRunFsPaths readerRun)) FS.ReadMode
+    readerCurrentEntryNo <- newPrimVar 0
+    -- load first page from disk, if it exists.
+    firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs bfs readerKOpsHandle
+    readerCurrentPage <- newIORef firstPage
+    return Reader {..}
+
+-- | This function should be called when discarding a 'Reader' before it
+-- was done (i.e. returned 'Empty'). This avoids leaking file handles.
+-- Once it has been called, do not use the 'Reader' any more!
+close ::
+     HasFS IO h
+  -> Reader (FS.Handle h)
+  -> IO ()
+close fs Reader {..} = do
+    FS.hClose fs readerKOpsHandle
+
+-- | The 'SerialisedKey' and 'SerialisedValue' point into the in-memory disk
+-- page. Keeping them alive will also prevent garbage collection of the 4k byte
+-- array, so they're long-lived, consider making a copy!
+data Result fhandle
+  = Empty
+  | ReadSmallEntry !SerialisedKey !(Entry SerialisedValue (BlobRef fhandle))
+    -- TODO: Do we need to return all of these fields?
+    -- TODO(optimise): Sometimes, reading the overflow pages is not necessary.
+    -- We could just return the page index and suffix length here to delay doing
+    -- the I\/O.
+  | ReadLargeEntry
+      !SerialisedKey
+      !(Entry SerialisedValue (BlobRef fhandle))
+      !RawPage
+      !Word32
+      ![RawOverflowPage]
+
+-- | Refers to the underlying 'Run', so make sure it remains open while reading
+-- blobs from it!
+data BlobRef fhandle = BlobRef !(Run.Run fhandle) !BlobSpan
+
+-- | Stop using the 'Reader' after getting 'Empty'!
+next ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> Reader (FS.Handle h)
+  -> IO (Result (FS.Handle h))
+next fs bfs reader@Reader {..} = do
+    entryNo <- readPrimVar readerCurrentEntryNo
+    page <- readIORef readerCurrentPage
+    go entryNo page
+  where
+    go !entryNo !page =
+        -- take entry from current page (resolve blob if necessary)
+        case rawPageIndex page entryNo of
+          IndexNotPresent -> do
+            -- if it is past the last one, load a new page from disk, try again
+            readDiskPage fs bfs readerKOpsHandle >>= \case
+              Nothing -> do
+                close fs reader
+                return Empty
+              Just newPage -> do
+                writeIORef readerCurrentPage newPage
+                writePrimVar readerCurrentEntryNo 0
+                go 0 newPage  -- try again on the new page
+          IndexEntry key entry -> do
+            modifyPrimVar readerCurrentEntryNo (+1)
+            let entry' = fmap (BlobRef readerRun) entry
+            return (ReadSmallEntry key entry')
+          IndexEntryOverflow key entry lenSuffix -> do
+            -- TODO: we know that we need the next page, could already load?
```

Comment:
If we wait for page to be loaded, it's probably not really faster. And I don't think prefetching does much, since we linear access pattern should already be predictable by the OS.
But we can see whether it's worth the (small) complexity once we can benchmark it.